In [22]:
import pandas as pd
import numpy as np
import scipy.stats as st

One tailed t-test - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

In [11]:
machine_one_tailed = pd.read_csv("files_for_lab/machine.txt", sep="\t", encoding="utf-16")

In [12]:
machine_one_tailed

New machine  Old machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [46]:
np.round(machine_one_tailed["Old machine"].mean(),2)

43.23

In [45]:
machine_one_tailed["New machine"].mean()

42.14

$$H0: \mu < 43.23$$
$$H1: \mu ≥ 43.23$$

In [47]:
old_machine_mean = machine_one_tailed["Old machine"].mean()
old_machine_std = machine_one_tailed["Old machine"].std(ddof=1)

new_machine_mean = machine_one_tailed["New machine"].mean()
new_machine_std = machine_one_tailed["New machine"].std(ddof=1)

sem = new_machine_std/np.sqrt(len(machine_one_tailed))


In [44]:
print("Old machine mean is: {:.2f}".format(old_machine_mean))
print("Old machine standard deviation is: {:.2f}".format(old_machine_std))
print()
print("New machin mean is: {:.2f}".format(new_machine_mean))
print("New machine standard deviation is: {:.2f}".format(new_machine_std))
print("The sem is: {:.2f}".format(sem))
print()
t = (new_machine_mean - old_machine_mean) / sem
print("Our statistic is: {:.2f}".format(t))
print("The p_value corresponding to our statistic is: {:.2f}".format(st.t.cdf(t,df = len(machine_one_tailed)-1)))
print("The significance level is set to 0.05")
print("We accept the H0?")

Old machine mean is: 43.23
Old machine standard deviation is: 0.75

New machin mean is: 42.14
New machine standard deviation is: 0.68
The sem is: 0.22

Our statistic is: -5.04
The p_value corresponding to our statistic is: 0.00
The significance level is set to 0.05
We accept the H0?


In [41]:
st.t.ppf(0.05,df = len(machine_one_tailed))

-1.8124611228107341

In [42]:
st.t.cdf(-5.04, df = len(machine_one_tailed))

0.000253313776858357

In [48]:
t_statistic, p_value = st.ttest_1samp(machine_one_tailed["New machine"], popmean = old_machine_mean, alternative = "less")
print("The t statistic of our sample is: {:.2f} and the corresponding p-value is: {:.2f}".format(t_statistic, p_value))

The t statistic of our sample is: -5.04 and the corresponding p-value is: 0.00


Matched Pairs Test - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [50]:
pokemon_matched_pairs = pd.read_csv("files_for_lab/pokemon.csv")

In [51]:
pokemon_matched_pairs

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

$$H0: \mu_{Attack} = \mu_{Defense} → \mu_{Attack} - \mu_{Defense} = 0$$
$$H1: \mu_{Attack} \neq \mu_{Defense} → \mu_{Attack} - \mu_{Defense} \neq 0$$

In [63]:
sample = pokemon_matched_pairs[["Attack", "Defense"]].sample(30, random_state = 1)
sample['Difference'] = np.abs(sample['Attack']-sample['Defense'])
sample.head(30)

Attack  Defense  Difference
8       104       78          26
510      92       75          17
175      46       34          12
735      50       58           8
242     105       75          30
764      38       33           5
540     120      120           0
430      70      160          90
84      100       70          30
642      87       63          24
255      35       35           0
111      95       85          10
289      35       55          20
349     140       70          70
703     100       80          20
521     110      130          20
437     104       71          33
565      60       45          15
224     125      230         105
371      48       43           5
301      30       30           0
537      75      130          55
341      40       50          10
135      95       57          38
708     125       90          35
47       80       70          10
666      55       55           0
60       82       78           4
741     100       62          38
117      65       95          30

In [64]:
sample_diff_mean, sample_diff_std = sample['Difference'].mean(), sample['Difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(25.333333333333332, 25.756597190828113)

In [65]:
t = sample_diff_mean / ( sample_diff_std / np.sqrt(len(sample)) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 25.33
The standard deviation of our samples differences is: 25.76
Our t statistics is: 5.39


In [68]:
tc = st.t.ppf(1-(0.05/2),df= len(sample))
np.round(tc, 2)

2.04

Our statistic is 5.39 while the critical value is 2.04. Then, as 5.39 > 2.04 we reject the H0. 

In [70]:
np.round(st.t.cdf(t,df = len(sample)), 6)

0.999996

The probability to see a t value as big as this one is for the t-Student distribution is:

$$P(x <=t=5.39) = 0.999996 > 0.025$$

we reject the H0.

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder

In [79]:
anova_laser = pd.read_excel("files_for_lab/anova_lab_data.xlsx")

In [86]:
anova_laser. columns = ["Power", "Etching Rate"]

In [89]:
anova_laser

Power  Etching Rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

- State the null hypothesis
$$H0: \mu_{1}=\mu_{2}=\mu_{3}=⋯=\mu_{n}$$
$$H1: \mu_{i}\neq\mu_{j}$$

for any i,j

H0: Changing the power has no effect on the etching rate

- State the alternate hypothesis

H1: Changig the power has an effect on the etching rate

- What is the significance level

0.05

- What are the degrees of freedom of model, error terms, and total DoF

In [97]:
d1 = anova_laser['Power'].nunique() - 1
d2 = len(anova_laser) - anova_laser['Power'].nunique()

print("Number of degrees of freedom d1: ",d1)
print("Number of degrees of freedom d2: ",d2)

Number of degrees of freedom d1:  2
Number of degrees of freedom d2:  12


In [87]:
group_df = anova_laser.groupby(["Power"])["Etching Rate"].agg(Etching_mean='mean',Samples='size').reset_index()

In [88]:
group_df

Power  Etching_mean  Samples
0  160 W         5.792        5
1  180 W         6.238        5
2  200 W         8.318        5

In [95]:
print(st.f_oneway(anova_laser[anova_laser["Power"] == "160 W"]["Etching Rate"], anova_laser[anova_laser["Power"] == "180 W"]["Etching Rate"], anova_laser[anova_laser["Power"] == "200 W"]["Etching Rate"] ))

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)


the p-values < 0.05, so we reject the H0